tests/integration/__init__.py

```diff
@@ -389,6 +394,51 @@ def _id_and_mark(f):
 requires_custom_payload = pytest.mark.skipif(SCYLLA_VERSION is not None or PROTOCOL_VERSION < 4,
                                             reason='Scylla does not support custom payloads. Cassandra requires native protocol v4.0+')
 xfail_scylla = lambda reason, *args, **kwargs: pytest.mark.xfail(SCYLLA_VERSION is not None, reason=reason, *args, **kwargs)
+
+
+def xfail_scylla_version_lt(reason, oss_scylla_version, ent_scylla_version, *args, **kwargs):
+    """
+    It is used to mark tests that are going to fail on certain scylla versions.
+
+    ent_scylla_version should contain at least first major: `x.1.1` or `x.1` or `x`
+    """
+    if not reason.startswith("scylladb/scylladb#"):
+        raise ValueError('reason should start with scylladb/scylladb#<issue-id> to reference issue in scylla repo')
+
+    if not isinstance(ent_scylla_version, (tuple, list)):
+        ent_scylla_version = (ent_scylla_version,)
+    first_major = None
+    ent_scylla_version = [Version(v) for v in ent_scylla_version]
+
+    # Enterprise releases are tricky, here are the rules:
+    # 1. If something is fixed in `a.b.c` it is also fixed in `a.b.d` where `c <= d`
+    # 2. If something is fixed in `a.1.1` it is also fixed in `c.d.f` where `Version(a.1.1)` <= `Version(c.d.f`)`
+    # To make version matching work properly of enterprise we need to have a first major version `x.1.1` where it was fixed.
+    for ent_v in ent_scylla_version:
+        if ent_v.minor in (1,0) and ent_v.micro in (1,0):
+            first_major = ent_v
+
+    if not first_major:
+        raise ValueError("ent_scylla_version should contain first major LTS: `x.1.1`")
```

Comment:
Note: can you put a link to documentation about enterprise versioning in the comment? If it is so weird, then I think there is a high risk of future change (accidental or intentional), so it would be good to have a source which we can follow in that case.

So the way I understand this comment now, is that "X.1.1" is somehow special.

If a something is fixed in "2024.0.1" then it may, or may not, be fixed in "2024.1.0". Reason: minor version differ, so first rule doesn't apply. None of the versions ends in "1.1" so second rule doesn't apply. This seems a bit weird.

If something is fixed in "2024.1.1" then it is also fixed in "2025.0.0" - second rule applies. Is that always the case?

Code: this function currently requires ent_scylla_version to either:
- Be a major version string (like "2024")
- Be a major + minor version string, if minor is 0 or 1 (like "2024.0" or "2024.1", but not "2024.2")
- Be a major + minor + patch string, if both minor and patch are 0 or 1 ("2024.1.1" or "2024.0.1", but not "2024.1.2").
- Be tuple, in which case first element of the tuple, when parsed as `Version`, must have major and minor in (0, 1). So `('2024', '2', '1')` is correct here, and so is `('2024.1', '2021.4')` - although I'm not sure what it would mean here and the function does not explain in any way what it expects. `('2024.2')` is incorrect.

Is that the intended behavior? Can you explain in the doc comment what exaclty this function expects from the arguments? It will be used in the future by other developers so it should be easily understandable and clear.

Reply:
> Note: can you put a link to documentation about enterprise versioning in the comment? If it is so weird, then I think there is a high risk of future change (accidental or intentional), so it would be good to have a source which we can follow in that case.

Unfortunately we don't document anywhere our enterprise versioning.
> So the way I understand this comment now, is that "X.1.1" is somehow special.
> 
> If a something is fixed in "2024.0.1" then it may, or may not, be fixed in "2024.1.0". Reason: minor version differ, so first rule doesn't apply. None of the versions ends in "1.1" so second rule doesn't apply. This seems a bit weird.

We don't have `0` in patch, but for sake of supporting `2025` or `2025.1` logic treats `2025.0` and `2025.0.0` the same way as `2025.1.1`.
It is not documented since it is something that we don't want to see in the code.

> If something is fixed in "2024.1.1" then it is also fixed in "2025.0.0" - second rule applies. Is that always the case?

We don't have `0` in patch, but yes, it is true.

> 
> Code: this function currently requires ent_scylla_version to either:
> 
> * Be a major version string (like "2024")
> * Be a major + minor version string, if minor is 0 or 1 (like "2024.0" or "2024.1", but not "2024.2")
> * Be a major + minor + patch string, if both minor and patch are 0 or 1 ("2024.1.1" or "2024.0.1", but not "2024.1.2").

True, but again, we don't want to see `0` in patch, and therefore do not document it.
> * Be tuple, in which case first element of the tuple, when parsed as `Version`, must have major and minor in (0, 1). So `('2024', '2', '1')` is correct here, and so is `('2024.1', '2021.4')` - although I'm not sure what it would mean here and the function does not explain in any way what it expects. `('2024.2')` is incorrect.

I have added a guard to ensure that every version in `ent_scylla_version` is an enterprise version.
I have also added a text to docstring that gives more context and explanation on what to expect.

> Is that the intended behavior? Can you explain in the doc comment what exaclty this function expects from the arguments? It will be used in the future by other developers so it should be easily understandable and clear.

I have put intended behavior into the docs, please take a look.

Reply:
@Lorak-mmk , I have simplified to support only string for `ent_scylla_version`